src/shogun/base/SGObject.cpp

```diff
@@ -1016,3 +939,57 @@ std::string CSGObject::to_string() const
 	ss << ")";
 	return ss.str();
 }
+
+bool CSGObject::equals(CSGObject* other, float64_t accuracy, bool tolerant)
+{
+	if (other==this)
+		return true;
+
+	if (other==nullptr)
+	{
+		SG_DEBUG("No object to compare to provided.\n");
+		return false;
+	}
+
+	/* Assumption: can use SGObject::get_name to distinguish types */
+	if (strcmp(this->get_name(), other->get_name()))
+	{
+		SG_DEBUG("Own type %s differs from provided %s.\n",
+				get_name(), other->get_name());
+		return false;
+	}
+
+	/* Assumption: objects of same type have same set of tags. */
+	for (auto it=self->map.cbegin(); it!=self->map.cend(); it++)
+	{
+		auto tag = it->first;
+		auto own = it->second;
+		auto given = other->get_parameter(tag);
+
+		SG_SDEBUG("Comparing parameter %s::%s.\n", this->get_name(),
+				tag.name().c_str());
+		if (own != given)
+		{
+			if (io->get_loglevel()<=MSG_DEBUG)
+			{
+				std::stringstream ss;
+				std::unique_ptr<AnyVisitor> visitor(new ToStringVisitor(&ss));
+
+				ss << "Own parameter " << this->get_name() << "::" << tag.name() << "=";
```

Comment:
I would actually like to print the differences here (also for vectors which is currently shown as `...`), as that was useful in unit-testing in the past

Reply:
It could be a custom visitor. Or we can add another interface 'Reducer'.